### ISSUE TITLE
Site Editor: add a + button that allows you to add any entity in a dropdown

### ISSUE THREAD

**Issue by annezazu** [WordPress/gutenberg#63378]
Currently, when you're navigating in the Site Editor, you are shown an "add new" in each section but you can only add a specific entity when you're in each section as shown below:

https://github.com/WordPress/gutenberg/assets/26996883/9e99aa1a-3b15-4f2a-9bd3-374de988d66b

Taking inspiration from the overall toolbar when viewing your site, let's consider adding something similar to save folks time and clicks:

![image (44)](https://github.com/WordPress/gutenberg/assets/26996883/82957953-79e6-4201-b074-d1e2a5badb85)

Mockup:

<img width="3380" alt="i5, dropdowns near term" src="https://github.com/user-attachments/assets/036eac3f-7457-457a-8ac9-b499d96005db">

Open questions:

- Should the menu should show contextual suggestions (surface as the top option the section you're already in) or have a set menu that doesn't change based on where you are?
- Should it show as + or + new to match what's shown on the front end? 

Those questions do not preclude adding the menu in the first place, as it'll be possible to adjust based on feedback.

See also [the comments](https://github.com/WordPress/gutenberg/issues/63378#issuecomment-2301697645) for some conversations around longer term possibilities.

**Comment by jasmussen** [comment]
Here's a quick sketch:

<img width="3380" alt="add new" src="https://github.com/user-attachments/assets/88173e5d-b28b-44b6-b695-683efa084fa4">

[Figma](https://www.figma.com/design/Uo7yuDG2EO8ssIm4VgFoGG/%22Add%22-%2F-%2B-Button?node-id=1-7938&t=z7Foo52gwg6HLy4M-1).

Since individual sections have their own dedicated "Add [item]" buttons with prominence, anything beyond that would fall into the quick tool category, hence the idea of the dropdown.

**Comment by jasmussen** [comment]
@WordPress/gutenberg-design any objections to me updating this issue and adding "needs dev"?

**Comment by jameskoster** [comment]
I like this idea in general but I'm not 100% convinced about the options presented in the mockup. 

I see they match the admin bar, but adding media items and users from within the site editor feels a bit off to me. Similarly I think I'd expect to find an option to create a pattern here.

It might be helpful to think about how this menu would work as a part of the broader admin design. Should it be contextual, or always present the same options? At the moment it feels like it would be contextual, which could be what is throwing me off.

**Comment by jasmussen** [comment]
Great feedback. Here's a fresh take that includes a plugin, as well as the frontend adminbar. Both with speculative visuals, but mainly to map out the behavior to ensure the path can lead in the right direction, rather than specifically suggesting any one or other path forward there. Dropdowns across the main dashboard, the site editor (aka "Design"), and a plugin, WooCommerce:

<img width="3380" alt="i3, dropdown menus" src="https://github.com/user-attachments/assets/1043c63b-6e7e-4044-ab4b-eaf01cb83ed0">

One idea you and I discussed in chat was also to leverage the command palette for these, pre-filled with "add". Worth including, even if this might not be where we start:

<img width="3380" alt="i3, command palette" src="https://github.com/user-attachments/assets/f12ca5fe-e5a6-4004-ae1e-d242fee41c71">

Let me know your thoughts, @WordPress/gutenberg-design! 

[Figma](https://www.figma.com/design/Uo7yuDG2EO8ssIm4VgFoGG/%22Add%22-%2F-%2B-Button?node-id=34-4111&t=H0qGrecL21UgGswC-1).

**Comment by jameskoster** [comment]
With the command palette approach, I reckon it should still return all options, but they could be ordered by context.

A drawback to the contextual menu is that it puts legitimate flows unreasonably out of reach. E.g: I just fulfilled an order, now I want to add a new page.

The command palette would solve this, but I understand it may not be feasible for a 'v1'. Alternative options to explore:

* The + menu is consistent regardless of context, containing all 'add' actions organised with flyouts. Probably needs a mockup as it could be unwieldy in practise. 
* There's a "More" flyout at the bottom of the menu containing all the other add actions ordered alphabetically.
* Show contextual actions, but include a search to locate other actions.

**Comment by jasmussen** [comment]
> With the command palette approach, I reckon it should still return all options, but they could be ordered by context.

Do we have a max amount of items to show? This is looking a bit overwhelming. Admittedly this example includes a plugin that adds additional CPTs, but still:

<img width="3380" alt="i3, command palette" src="https://github.com/user-attachments/assets/83e4fcda-614b-4177-a00e-c5a608cf4685">

**Comment by jasmussen** [comment]
I'm not really feeling the consistent-everywhere dropdown:

![Screenshot 2024-08-22 at 10 07 25](https://github.com/user-attachments/assets/e0da2e95-4b6b-4c15-80df-55a337d250a5)

Feel free to tinker if I'm misconstruing your idea!

For now I'll try to update the original issue with both a reduced mockup just for the site editor.

**Comment by jasmussen** [comment]
I've created a minimal MVP mockup that simply replicates the dropdown menu in the site editor. This will have intrinsic value in saving clicks, and solves the original issue. The conversation around the longer term view can likely also be more informed based on what we learn from that V1, i.e. whether it's the command palette or something else; the conversation here and mockups will be there for the V2.

**Comment by ntsekouras** [comment]
I don't believe there is a way to know who registered a custom post type through [register_post_type](https://developer.wordpress.org/reference/functions/register_post_type/). My comment refers to the mockup with `WooCommerce` having a nested dropdown.

**Comment by jameskoster** [comment]
> This is looking a bit overwhelming.

I don't mind it, to be honest. It's fairly common to see a lot of suggestions in cmd+k interfaces. To refine the list you can type to search for whatever you're trying to add.

We might follow-up to show contextual suggestions by default, and 'hide' others behind a "More" button:

<img width="1440" alt="palette" src="https://github.com/user-attachments/assets/1b6b93ae-6913-48b6-b6fb-1eaf051e924c">

But I don't know if it's worth the additional logic. Ordering by context and leaning of search seems simpler to me.

**Comment by richtabor** [comment]
Are the icons necessary, and the "Add new" label up top? 

Is the idea that this emulates the + New from the dashboard, or a different set of items to add?

**Comment by jasmussen** [comment]
> Are the icons necessary, and the "Add new" label up top?

Seems okay without it too. No strong opinion:

<img width="3380" alt="i5, dropdowns near term" src="https://github.com/user-attachments/assets/85ebc166-e95b-45d3-857d-734a768dfa0b">

> Is the idea that this emulates the + New from the dashboard, or a different set of items to add?

Yes, the idea is that this "+ New" button available in the adminbar become available in these immersive situations. Whether we always show a consistent/full set of options, or a subset (contextual or otherwise), I think we can explore.
